guardian/service-catalogue
# feat(cloudquery): Run on ECS

## What does this change? And why?
Inspired by [our experiments](https://github.com/guardian/container-experiments) during this quarter's R&D sprint, this change moves away from using EC2 to run CloudQuery and towards ECS, on Fargate.

This brings some benefits:
- No idle time
  
  The EC2 instance sits idle for ~23 hours a day.
- Implementing different schedules is easier
  
  Though it's possible to implement different schedules using systemd timers on EC2, we'd likely need to manage multiple config files, which isn't fun.
  
  The changes in this PR allow us to schedule a job as follows:

  ```ts
  const customRateTables: CustomRateTable[] = [
    {
      schedule: Schedule.rate(Duration.hours(2)),
      tables: ['aws_s3_buckets'],
    },
    {
      schedule: Schedule.rate(Duration.minutes(30)),
      tables: ['aws_lambda_functions'],
    },
  ];
  ```

  Being able to collect data from different tables at different rates is going to be crucial to adding GitHub to CloudQuery due to the rate limiting of the GitHub API. If we collect data for [`github_teams`](https://www.cloudquery.io/docs/plugins/sources/github/tables/github_teams) at a slower rate than [`github_repositories`](https://www.cloudquery.io/docs/plugins/sources/github/tables/github_repositories), we reduce the chance of getting rate limited.

- Removal of known errors

  The EC2 collector currently produces error logs when collecting data for the [organization tables](https://www.cloudquery.io/docs/plugins/sources/aws/tables/aws_organizations) . This is because we're making the request from all accounts. We know that only the DeployTools account has [this permission](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_info-operations.html), so we now make the request from just that account.
  
  Similar to the above, this is possible when using EC2, however we'd need to manage multiple config files, which isn't fun.

- Reduced infrastructure

  With the EC2 version, we have an ASG purely to facilitate deployments via Riff-Raff. This is not needed in the ECS version.

  In the EC2 version, we also place configuration in S3. The implementation of the ECS version removes this, in favour of inlining the config.

## How has it been verified?
The stack has been [deployed](https://riffraff.gutools.co.uk/deployment/view/d560b09d-901d-4863-8260-bfbded79a4db). There are a couple of ways to verify it works:
- Observe the [logs in Central ELK](https://logs.gutools.co.uk/goto/6c10fdf0-f6e5-11ed-a27a-7333ae0cfb0a)
- View the configuration of the ECS Cluster in AWS

## Notes
Sadly, I don't think it will be possible to use [centralised logging container](https://aws.amazon.com/blogs/opensource/centralized-container-logging-fluent-bit/) on Fargate as:

> Fargate tasks do not support the `DAEMON` scheduling strategy.
> – https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_services.html#service_scheduler

This means each task has three containers:
1. One to vend temporary AWS RDS IAM tokens
2. One to perform the CloudQuery work
3. One to ship logs

We should keep an eye on the cost of this, as it might become cheaper to ship the logs to CloudWatch logs, and then use [cloudwatch-logs-management](https://github.com/guardian/cloudwatch-logs-management) to route them to Central ELK.

## TODO (non-blocking)
- Fine tuning, documentation (and renaming) of https://github.com/guardian/hackday-firelens, as the logs lines in Central ELK do not quite match those from the EC2 version, with missing fields.
